core/src/main/scala/org/http4s/RequestInfo.scala

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2013 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s
+
+import cats._
+import cats.syntax.all._
+import org.http4s.internal.reduceComparisons
+
+/** A projection of a [[Request]] without the body.
+  *
+  * @note The [[Request#attributes]] are omitted in this encoding because they
+  *       do not (and can not) have a [[cats.kernel.Order]] instance. If they
+  *       were included here, then we could not write a [[cats.kernel.Order]]
+  *       instance for [[RequestInfo]], limiting some of its utility, e.g. it
+  *       could not be used in a [[cats.data.NonEmptySet]].
+  */
+sealed abstract class RequestInfo extends Product with Serializable {
+  def headers: Headers
+  def httpVersion: HttpVersion
+  def method: Method
+  def uri: Uri
+
+  def withHeaders(value: Headers): RequestInfo
+  def withHttpVersion(value: HttpVersion): RequestInfo
+  def withMethod(value: Method): RequestInfo
+  def withUri(value: Uri): RequestInfo
```

Comment:
Do you actually want to manipulate these, or just project them from an existing request?

Reply:
I can see a lot of use cases for manipulating them. Mostly relating to normalizing them for the purposes of logging, caching, and metrics. That is, a user might only care that the uri and method are the same, and thus they may,

```scala
_.withHeaders(Headers.empty).withHttpVersion(HttpVersion.`HTTP/1.1`)
```

I also just tend to add both `withFoo` and `mapFoo` methods as a rule these days to data types which don't expose a `copy` for binary compatibility. If we wanted to add `withHeaders` later, without breaking binary compatibility, we'd have to do some gymnastics.

```scala
sealed abstract class RequestInfo {
/* elided */

// Default to make MiMa not complain
  def withHeaders(value: Headers): RequestInfo = headers
}

object RequestInfo {
    private[this] final case class RequestInfoImpl(/* elided */) extends RequestInfo {
      override def withHeaders(value: Headers): RequestInfo = this.copy(headers = value)
    }
/* elided */
}
```